dlrover/python/master/monitor/metric_context.py

```diff
@@ -0,0 +1,317 @@
+    """
+
+    def __init__(
+        self,
+        npu_total_mem=0,
+        npu_used_mem=0,
+        npu_util=0,
+        npu_temperature=0,
+        npu_health_state=1,
+        npu_link_state=1,
+        npu_optical_state=1,
+        npu_network_state=1,
+        npu_tx=0.0,
+        npu_rx=0.0,
+    ):
+        super().__init__("ascend.NPU")
+        self.metrics = {
+            NpuMetricType.NPU_TOTAL_MEM: npu_total_mem,
+            NpuMetricType.NPU_USED_MEM: npu_used_mem,
+            NpuMetricType.NPU_UTIL: npu_util,
+            NpuMetricType.NPU_TEMP: npu_temperature,
+            NpuMetricType.NPU_HEALTH_STATE: npu_health_state,
+            NpuMetricType.NPU_LINK_STATE: npu_link_state,
+            NpuMetricType.NPU_OPTICAL_STATE: npu_optical_state,
+            NpuMetricType.NPU_NETWORK_STATE: npu_network_state,
+            NpuMetricType.NPU_RDMA_TX: npu_tx,
+            NpuMetricType.NPU_RDMA_RX: npu_rx,
+        }
+
+    def set_metric(self, key, value):
+        if key in self.metrics.keys():
+            self.metrics[key] = value
+
+    def get_metric(self, key):
+        if key in self.metrics.keys():
+            return self.metrics[key]
+        else:
+            return None
+
+
+class NodeXpuMetric(object):
+    """
+    Metrics of all XPUs in a single node
+
+    list of XpuMetric with index as local rank id
+    """
+
+    def __init__(self):
+        pass
+
+    @abstractmethod
+    def update_avg_metrics(self):
+        pass
+
+
+class NodeGpuMetric(NodeXpuMetric):
+    """
+    Metrics of all GPUs in a single node
+
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.node_metrics: Dict[int, GpuMetric] = {}
+        self.avg_metrics = GpuMetric()
+
+    def update_avg_metrics(self):
+        self.avg_metrics.metrics[GpuMetricType.GPU_FREE_MEM] = 0
+        self.avg_metrics.metrics[GpuMetricType.GPU_USED_MEM] = 0
+        self.avg_metrics.metrics[GpuMetricType.GPU_UTIL] = 0
+        self.avg_metrics.metrics[GpuMetricType.GPU_SM_UTIL] = 0.0
+        self.avg_metrics.metrics[GpuMetricType.GPU_TENSOR_UTIL] = 0.0
+
+        for _, metric in self.node_metrics.items():
+            self.avg_metrics.metrics[
+                GpuMetricType.GPU_FREE_MEM
+            ] += metric.get_metric(GpuMetricType.GPU_FREE_MEM)
+            self.avg_metrics.metrics[
+                GpuMetricType.GPU_USED_MEM
+            ] += metric.get_metric(GpuMetricType.GPU_USED_MEM)
+            self.avg_metrics.metrics[
+                GpuMetricType.GPU_UTIL
+            ] += metric.get_metric(GpuMetricType.GPU_UTIL)
+            self.avg_metrics.metrics[
+                GpuMetricType.GPU_SM_UTIL
+            ] += metric.get_metric(GpuMetricType.GPU_SM_UTIL)
+            self.avg_metrics.metrics[
+                GpuMetricType.GPU_TENSOR_UTIL
+            ] += metric.get_metric(GpuMetricType.GPU_TENSOR_UTIL)
+
+        self.avg_metrics.metrics[GpuMetricType.GPU_FREE_MEM] = round(
```

Comment:
Metrics are closely tied to time value. Aggregation results shouldn't be obtained based on the number of elements.

Reply:
U can consider using 'statsd'.